README.md
The 'DiffXTables' R package
===============================

[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/DiffXTables)](https://cran.r-project.org/package=DiffXTables)
[![CRAN_latest_release_date](https://www.r-pkg.org/badges/last-release/DiffXTables)](https://cran.r-project.org/package=DiffXTables)
[![metacran downloads](https://cranlogs.r-pkg.org/badges/DiffXTables)](https://cran.r-project.org/package=DiffXTables)
[![metacran downloads](https://cranlogs.r-pkg.org/badges/grand-total/DiffXTables)](https://cran.r-project.org/package=DiffXTables)



### Overview

Pattern heterogeneity between two variables across conditions is often fundamental to a scientific inquiry. For example, a biologist could ask whether co-expression between two genes in a cancer cell has been modified from a normal cell. The 'DiffXTables' R package answers such questions via evaluating statistical evidence for distributional changes in the involved variables based on observed data without using a parametric mathematical model.

The package provides statistical methods for hypothesis testing of differences in the underlying distributions across two or more contingency tables. They include five statistical tests: 

1. The comparative chi-squared test: Its statistical foundation is first established in [(Song et al. 2014) <doi:10.1093/nar/gku086>](https://doi.org/10.1093/nar/gku086). It is later extended to identify differential patterns in networks [(Zhang et al. 2015) <doi:10.1093/nar/gkv358>](https://doi.org/10.1093/nar/gkv358).
2. The Sharma-Song test: The test detects differential departure from independence via second-order difference in the joint distributions underlying two or more contingency tables. The test is fully described in [(Sharma et al. 2021) <doi:10.1093/bioinformatics/btab240>](https://doi.org/10.1093/bioinformatics/btab240).
3. The heterogeneity test: It is described in (Zar, 2010) and widely appears in textbooks. In contrast to the above comparative chi-squared test, it is not always powerful as demonstrated by examples in the package vignette.
4. The marginal-change test: It is determines the first-order (marginal) differences across conditions [(Sharma et al. 2020) <doi:10.1145/3388440.3412485>](https://doi.org/10.1145/3388440.3412485).
5. The strength test: It determines the strength of association of all the conditions [(Sharma et al. 2020) <doi:10.1145/3388440.3412485>](https://doi.org/10.1145/3388440.3412485).

The package also provides a comparative type analysis of difference in association across contingency tables to reveal the highest order of their differences. 

Their null test statistics all follow an asymptotically chi-squared null distribution. These options test for heterogeneous patterns that differ in either the first order (marginal) or the second order (joint distribution deviation from product of marginals). Second-order differences may reveal more fundamental changes than first-order differences across heterogeneous patterns.

### When to use the package

This package takes a model-free approach without assuming an underlying parametric model for the relationship between variables, in contrast to differential correlation based on differences between linear models. Its input is contingency tables that store the counts or frequencies of discrete variables. Thus, continuous variables need to be discretized before using the tests. One option to do discretization is via optimal univariate clustering provided by the ['Ckmeans.1d.dp'](https://cran.r-project.org/package=Ckmeans.1d.dp) R package.

### To download and install the package

```{r}
install.packages("DiffXTables")
```


NEWS
NEWS

Version 0.1.3

  2021-05-14
  1. Updated user manuals, README, DESCRIPTION, and the vignette.
  
  2021-04-28
  1. Added the citation to the Bioinformatics paper describing 
     Sharma-Song test with observed null table marginals in 
     DESCRIPTION, CITATION, REFERENCES.bib, and README.md files. 
  2. Introduced a boolean parameter to compensate the issue of Cochran's
     condition of samples value less than 5 in any cell.
  
  2020-12-24
  1. Updated DESCRIPTION
  2. Introduced null.table.marginal parameter to use the
     sharma.song.test with observed and uniform marginal null
     population.
  
  3. No longer using the correction of normalized frequency table A
     inside function get.e.mat()
  4. Added test cases for the uniform marginal null tables.
  5. Added test cases for parameter compensated parameter.
  6. Updated the user manual for sharma.song.test() function.
  7. Added full citation information and DOI number for the 
     first-order marginal change test ACM paper (Sharma et al, 2020)
     in DESCRIPTION, CITATION, REFERENCES.bib, and README.md files.
  8. stats::chisq.test() is no longer called inside function
     chisq.test.stat(). This reduce the overhead in chisq.test(), as we
     only need to compute the Pearson chi-squared test statistic.
  

  2020-12-16
  1. Created version 0.1.3 from 0.1.2

Version 0.1.2

  2020-08-16
  1. Created version 0.1.2 from 0.1.1
  2. Fixed a NOTE stating Rdpack package was imported but not used.
  
Version 0.1.1
  
  2020-08-11
  1. Removed DOI from reference (Sharma et al 2020), which is still
     in press.
  
  2020-08-07
  1. Updated DESCRIPTION
  2. Updated REFERENCES.bib and CITATION
  
  2020-08-02
  1. Updated user manual for type.analysis() function
  2. Turned off Yates correction in calculating 2x2 chi-squared
     statistics in internal R function chisq.test.stat
  3. Adjusted test cases for 2x2 tables used for the heterogeneity 
     test without Yates correction. See line 270 and 272 in 
     test_DiffXTables.R
  4. Removed duplicate function chisq.test.stat() and created a new
     file chisq-test-stat.R to include only one copy of the function.
  5. Changed the package title to "Pattern Analysis Across Contingency Tables"
  
  2020-08-01
  1. Renamed association.order.ana() to type.analysis()
  2. Included a S3 method print.DiffXTableTypeAnalysis()
  3. Updated the printout of type.analysis()
  
  2020-07-22
  1. Added a marginal.change.test for first-order analysis.
  2. Added a strength.test for determining the strength of
     association.
  3. Added a association.order.ana method for analysis of strength
     and change in association.
  4. Added test cases for all the methods. 

  2020-05-14
  1. Corrected the normalized frequency table A inside 
     function get.e.mat() in Sharma-Song.R when an entry 
     has both zero row and zero column sums.
  2. Add new test cases for the above change.
  
  2020-05-02
  1. Created version 0.1.1 from 0.1.0
  
Version 0.1.0 

  2020-03-19
  1. Changed the return value of the simulator function 
     SimulateDiffXTab().
  2. Renamed the simulator function to simulate_diff_tables() from
     SimulateDiffXTab().
  3. Minor editing in various files to improve code readability.
  4. More tests are added for the simulator function.
  5. Corrected typos in the vignette.

  2020-03-18
  1. Created version 0.1.0
  2. Removed unnecessary calculation of the rank of covariance of
     matrix C, which is theoretically known. Thus the import of Matrix
     package is removed.
  3. Revised function manuals.
  4. Changed the simulator function arguments.
  
Version 0.0.4 (not publicly released)

  2020-01-23
  1. Created version 0.0.4
  2. A simulator function SimulateDiffXTab() is added.
  
Version 0.0.3 (not publicly released)

  2019-11-12
  1. Created version 0.0.3.

Version 0.0.2

  2019-11-12
  1. Updated the package description.
  2. Updated the REAMDE.md file.

  2019-11-10
  1. Updated the vignette.
  2. Updated the REAMDE.md file.

  2019-10-29
  1. Created version 0.0.2.
  2. Fixed a bug for 2x2 tables in the Sharma-Song test. A test case
     is added to check for the bug.
  3. Changed package title to "Pattern Heterogeneity via
     Distributional Differences Across Contingency Tables"

Version 0.0.1

  2019-10-29
  1. Created version 0.0.1.

Version 0.99.1 (not publicly released)

  2019-10-23
  1. Created version 0.99.1.

Version 0.99.0 (not publicly released)

  2019-10-20
  1. Created version 0.99.0, the first version of the package.


tests/testthat.R
library(testthat)
library(DiffXTables)

test_check("DiffXTables")


R/strength-test.R
# strength-test.R
#
# Method to determine strength of association (zero-order) in each table.
# Input : a list of sampled matrices, must be greater or equal to 2
# Output: Statistics and p-value, significance of association strength.
# Created by: Ruby Sharma 
# Date Created: May 17 2020
# Updated: August 2, 2020. MS

#' @export
strength.test <- function(tables)
{
  K = length(tables)
  
  if (mode(tables) != "list" || K < 2) {
    stop("Input tables must be a list of two or more matrices!")
  }
  
  KStats <- vector("numeric", K)
  df <- vector("integer", K)
  
  for(i in seq(K)){
    if(sum(tables[[i]]>0)){
      KStats[i] <- chisq.test.stat(tables[[i]])
    }else{
      KStats[i] <- 0
    }
    df[i] <- prod(dim(tables[[i]]) - 1)
  }
  
  Stat <- sum(KStats)
  Df <- sum(df)
  
  P.val <- pchisq(Stat, Df, lower.tail = FALSE)
  
  names(Stat) <- "X-squared"
  names(Df) <- "df"
  DNAME <- deparse(substitute(tables))
  
  return(structure(list(
    statistic = Stat,
    parameter = Df,
    p.value = P.val,
    data.name = DNAME,
    method = "Strength Test for Association in Multiple Contingency Tables"),
    class = "htest"))
}


R/marginal-change-test.R
# marginal-change-test.R
#
# Method to determine first-order (marginal) differential patterns.
# Input : a list of sampled matrices, must be greater or equal to 2
# Output: Statistics and p-value, significance of marginal differentiality
# Created by: Ruby Sharma and Dr. Joe Song
# Date Created: May 17 2020
# Updated: August 2, 2020. MS. Turned off correction on 2x2 tables.

#' @importFrom  Rdpack reprompt

#' @export
marginal.change.test <- function(tables)
{
  K <- length(tables)
  if (mode(tables) != "list" || K < 2) {
    stop("Input must be a list of 2 or more matrices!")
  }
  
  for (k in seq(K-1)) {
    if (!identical(dim(tables[[k]]), dim(tables[[k + 1]]))) 
      stop("All matries should have same number of rows and columns")
  }
  
  ## Constructing row marginal table
  r = nrow(tables[[1]])
  
  Xc = matrix(nrow = K, ncol = r)
  for (i in seq(K)) {
    Xc[i, ] = rowSums(tables[[i]])
  }
  
  ## Constructing column marginal table
  s = ncol(tables[[1]])
  Yc = matrix(nrow = K, ncol = s) 
  for (i in seq(K)) {
    Yc[i, ] = colSums(tables[[i]])
  }
  
  ## Evaluating Chisq of all the tables 
  Xchisq = 0
  Ychisq = 0
  if(sum(Xc) > 0 || sum(Yc) > 0) {
    if(sum(Xc)!=0){
      Xchisq = chisq.test.stat(Xc)
    }
    if(sum(Yc)!=0){
      Ychisq = chisq.test.stat(Yc)
    }
    
    ## Determining the strength of marginal difference between the tables
    Stat = Xchisq+Ychisq	
    Df = (K-1)*((r+s)-2)
    P.val = pchisq(Stat , Df, lower.tail = FALSE)
  } else {
    Stat = 0	
    Df = 0
    P.val = 1
  }
  
  names(Stat) <- "X-squared"
  names(Df) <- "df"
  DNAME <- deparse(substitute(tables))
  
  return(structure(list(
    statistic = Stat,
    parameter = Df,
    p.value = P.val,
    data.name = DNAME,
    method = "Test for Marginal Change Across Contingency Tables"),
    class = "htest"))
  
}


R/CPChisq.R
# CPChisq.R -- comparative chi-square test for heterogeneous associations
#
# MS
# Created: July 3, 2015 generalized from two contingency tables in
#                       "hetero-chisq-steel.R"
# Modified: Feb 5, 2016. MS
#   Modified the method name
#   Added log.p argument to specify log p-value


#' @importFrom stats pchisq pnorm
#' @export
cp.chisq.test <- function(tables, method = c("chisq", "nchisq", "default", "normalized"), log.p=FALSE)
{
  DNAME <- deparse(substitute(tables))
  if(mode(tables)!="list" || length(tables)<2 )
  {
    stop("only accept list of 2 or more matrices as input!")
  }

  for(k in 1:(length(tables)-1)){
    if(!identical(dim(tables[[k]]), dim(tables[[k+1]])))
      stop("All matries should have same number of rows and columns")
  }

  method <- match.arg(method)
  if(method == "default") {
    warning(paste0("method=\"", method, "\" is deprecated. Use \"chisq\" instead."))
    method <- "chisq"
  } else if(method == "normalized") {
    warning(paste0("method=\"", method, "\" is deprecated. Use \"nchisq\" instead."))
    method <- "nchisq"
  }

  pooled <- tables[[1]]
  for(i in 2:length(tables)) {
    pooled <- pooled + tables[[i]]
  }

  # Row sum of matrix pooled
  pooled.row.sum <- apply(pooled, 1, sum)

  # Column sum of matrix pooled
  pooled.col.sum <- apply(pooled, 2, sum)

  #Sum of matrix pooled
  pooled.sum <- sum(pooled.col.sum)

  pooled.chisq <- 0

  if(pooled.sum > 0) {
    #Expected value in pooled
    pooled.expected <- pooled.row.sum %*% t(pooled.col.sum) / pooled.sum

    #Chi-square value of pooled
    pooled.chisq <- sum((pooled - pooled.expected)^2 / pooled.expected, na.rm = TRUE)

    total.chisq <- 0

    for(i in seq_along(tables)) {
      #Expected values in matrix i
      expected <- pooled.expected / pooled.sum * sum(tables[[i]])

      #Chi-square value of matrix i
      chisq <- sum((tables[[i]] - expected)^2 / expected, na.rm = TRUE)

      total.chisq <- total.chisq + chisq
    }

    hetero.chisq  <- total.chisq - pooled.chisq
    df <- (length(tables) - 1) * (sum(pooled.row.sum != 0) - 1) *
      (sum(pooled.col.sum != 0) - 1)

  } else {

    hetero.chisq <- 0
    df <- 0

  }

  if(method=="chisq") {

    finalStat <- hetero.chisq
    names(finalStat) <- "X-squared"

    finalDf <- df
    names(finalDf) <- "df"

    p.val <- pchisq(finalStat, df = finalDf, lower.tail=FALSE, log.p=log.p)
    names(p.val) <- "p.value"

    return(structure(list( statistic=finalStat, parameter=finalDf, p.value=p.val,
                           method = "Comparative Chi-Squared Test for Pattern Difference",
                           data.name= DNAME),
                     class = "htest"))

  } else if(method=="nchisq") {

    finalStat <- ifelse(df > 0, ( hetero.chisq - df) / sqrt( 2 * df ), 0)
    names(finalStat) <- "Normalized X-squared"

    finalDf <- df
    names(finalDf) <- "df"

    p.val <- ifelse(df > 0, pnorm( finalStat, lower.tail=FALSE, log.p=log.p), 1)
    names(p.val) <- "p.value"

    return(structure(list(statistic = finalStat, parameter = finalDf, p.value = p.val,
                          method = "Nomalized Comparative Chi-Squared Test for Pattern Difference",
                          data.name= DNAME),
                     class = "htest"))
  }
}


R/heterogeneity.R
# heterogeneity.R -- heterogeneity chi square test
# Created by: Ruby Sharma
# Created: January 7, 2019
# Modified: Feb 5, 2016. MS

#' @export
heterogeneity.test <- function(tables)
{

   if(mode(tables)!="list" || length(tables)<2 )
   {
      stop("only accept list of 2 or more matrices as input!")
   }

   for(k in 1:(length(tables)-1)){
      if(!identical(dim(tables[[k]]), dim(tables[[k+1]])))
        stop("All matries should have same number of rows and columns")
   }

   Hp <- Reduce('+', tables)
   if(all(Hp == 0))
   {
     Stat <- 0
      df <- 0
   }else{
      xall <-  unlist(lapply(seq_along(tables),  function(k){

               return(chisq.test.stat(tables[[k]]))
       }))
       xHp <- chisq.test.stat(Hp)

       # heterogeneity chi square, summation of chisq - pooled chisq
       Stat <- sum(xall) -xHp
       df <- (length(tables)* prod(dim(Hp) - 1)) - prod(dim(Hp) - 1)
  }

  p.val = pchisq(abs( Stat), df, lower.tail=F)
  DNAME <- deparse(substitute(tables))
  
  names( Stat) <- "X-squared"
  names(df) <- "df"
  
  return(structure(list(
    statistic = abs(Stat), 
    parameter = df, 
    p.value = p.val,
    data.name = DNAME,
    method = "Heterogeneity Test for Pattern Difference"),
    class = "htest"))

}




R/type-analysis.R
# type-analysis.R
#
# Method to determine Type-0, Type-1 and Type-2 patterns in the set of 
# contingency tables.
# Input : a list of sampled matrices, must be greater or equal to 2
# Output: Statistics and p-value, significance of zero-order, first-order and
#         second-order test.
# Created by: Ruby Sharma and Dr. Joe Song
# Date Created: July 20 2020
# Updated: August 2, 2020 MS

#' @importFrom  pander pandoc.emphasis
#' @importFrom  pander pander

#' @export
type.analysis <- function(tables, alpha = 0.05)
{
  K <- length(tables)
  if (mode(tables) != "list" || K < 2) {
    stop("Must input a list of two or more matrices!")
  }
  
  for (k in seq(K-1)) {
    if (!identical(dim(tables[[k]]), dim(tables[[k + 1]]))) 
      stop("All matries must have the same dimension!")
  }
  
  SecondOrd <- sharma.song.test(tables)
  MarginalChange <- marginal.change.test(tables)
  StrengthCond <- strength.test(tables)
  
  Type <- decide.type(
    StrengthCond$p.value, MarginalChange$p.value, SecondOrd$p.value, 
    alpha)
  
  return(structure(list(
    Zeroth.order = StrengthCond,
    First.order = MarginalChange,
    Second.order = SecondOrd,
    Type = Type,
    Method = "Type Analysis Across Multiple Contingency Tables"),
    class = "DiffXTableTypeAnalysis"
  ))
}

#' @export
print.DiffXTableTypeAnalysis <- function(x, ...)
{
  # Printing differential table type analysis 
  
  with(x, {
    # TypeNull <- "X and Y are statistically independent in all K conditions
            # and same X marginal distribution across conditions so does Y."
    
    # Constructing Result table
    Result = data.frame(matrix(nrow = 3, ncol = 3))
    Result[1,] = c(
      '0th', AddStars(Zeroth.order$p.value), 
      "association present in tables" )
    Result[2,] = c(
      '1st', AddStars(First.order$p.value), 
      "differential in marginal distribution across tables" )
    Result[3,] = c(
      '2nd', AddStars(Second.order$p.value), 
      "differential in deviation from joint distribution to product of marginals" )
    
    colnames(Result) = c("Order", "P value", "Description")
    
    pander("\nComprehensive Type Analysis of Multiple Contingency Tables:\n\n")
    
    if(is.null(Type)) {
      pander("\tType null: association absent")
    } else if(Type == 0) {
      pander("\tType 0: association present; patterns conserved")
    } else if(Type == 1) {
      pander("\tType 1: association present; difference up to first order")
    } else if(Type == 2) {
      pander("\tType 2: association present; difference up to second order")
    }
    
    pander(Result, style = "grid")
    pander("Signif. codes: 0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1" )
    # cat("\n\n")
    # pandoc.emphasis(Message)
  } )
  invisible(x)
}

# Deciding the differential type based on order 0, 1, and 2 p.value
decide.type <- function(
  Active.P.value, Marginal.P.value, SecondOrd.P.value, alpha
)
{
  if(Active.P.value <= alpha && 
     Marginal.P.value > alpha && 
     SecondOrd.P.value > alpha){

    Type <- 0
  } else if(Active.P.value <= alpha && 
            Marginal.P.value <= alpha && 
            SecondOrd.P.value > alpha){
    Type <- 1
  } else if(Active.P.value <= alpha && 
            SecondOrd.P.value <= alpha){
    Type <- 2

  } else {
    Type <- NULL
  }
  
  return(Type)
}

## Assigning significance stars to p.value
AddStars <- function(P.val)
{
  if(P.val > 0.1){
    stars = " "
  } else if(P.val > 0.05 && P.val <= 0.1 ){
    stars = "."
  } else if(P.val > 0.01 && P.val <= 0.05){
    stars = "*"
  } else if(P.val > 0.001  && P.val <= 0.01){
    stars = "**"
  } else {
    stars = "***"
  }
  P.val <- signif(P.val, digits = 5)
  P.val <- paste0(P.val, " ", stars) 
  return(P.val)
}



R/chisq-test-stat.R
# chisq-test-stat.R
#
# Extracted from heterogeneity.R and marginal-change-test.R 
#   to compute the chi-square statistic only 

# #' @importFrom stats chisq.test

# Calculate chi-square test after removing zero rows or columns
#chisq.test.stat <- function(table)
#{
#  non.zero.rows <- apply(table, 1, function(row) { 0 != sum(abs(row)) } )
#  non.zero.cols <- apply(table, 2, function(col) { 0 != sum(abs(col)) } )

#  perform Pearson chi-square test
#  table = table[non.zero.rows, non.zero.cols]
#  if(all(dim(table) > 1)) { # tables needs to have more than one row and one column
#    if(length(table) > 1) {
#    chisq <- suppressWarnings(
#      chisq.test(table, correct = FALSE)$statistic)
#  } else {
#    # stop("'Table' must have at least 2 elements")
#    chisq <- 0
#  }
  
#  return(chisq)
#}

# Calculate chi-square statistics
chisq.test.stat <- function(table){
  
  if(all(dim(table) > 1)) {
    nr <- nrow(table)
    nc <- ncol(table)
    sr <- rowSums(table)
    sc <- colSums(table)
    n <- sum(sr)
    E <- outer(sr, sc, "*")/n
    chisq <- sum((table-E)^2/E, na.rm=TRUE)
  } else {
    stop("'table' must be least 2x2")
    chisq <- 0
  }
  return(chisq)
}


R/Sharma-Song.R
# Method to determine second order (interaction) differential patterns
# Input : a list of sampled matrices, must be greater or equal to 2
# Output: Statistics and p-value, significance of differentiality
# Created by: Ruby Sharma and Dr. Joe Song
# Date Created: 28 December 2018
# Modified: 
#   12 June 2019 by Dr. Joe Song
# 
#   October 29, 2019:
#     - Added ifelse to check if U is a matrix, if, yes, multiply by 
#       nrow(U) else multiply by 1
#     - Added a test case for 2x2 tables
#     - Renamed matrix U to Q to be consistent with the manuscript.
#     - Simplified the calculation of Df.
# 
#   March 18, 2020: The rank of the covariance matrix is always
#     K-1. No need to compute the rank numerically.
#
#   May 14, 2020: Corrected the normalized frequency table A inside 
#     function get.e.mat() when an entry has both zero row and zero 
#     column sums.
#
#   December 17, 2020: 
#   Introduced a new parameter to select the marginal distribution of the 
#   tables in null population.
#   Two options for null.table.marginal:
#   observed: Utilizes the observed marginal in null population to determine the
#   second order differential patterns [default]
#   uniform: Utilizes the uniform marginal in null population to determine the
#   second order differential patterns.
#   
#   April 20, 2021: 
#   Introduced a boolean parameter 'compensated' to avoid the issue of upwards bias 
#   originating due to small expected value (Cochran's condition). Default is FALSE.
  

#######################################
#### Imports for the whole package ####
# No longer need this import 
#XXXXXXXXXX' @importFrom Matrix rankMatrix
#######################################

#' @export
sharma.song.test <- function(tables, null.table.marginal = c("observed", "uniform"),
                             compensated = FALSE)
{
  if (mode(tables) != "list" || length(tables) < 2) {
    stop("Input must be a list of 2 or more matrices!")
  }

  for(k in 1:(length(tables)-1)){
    if(!identical(dim(tables[[k]]), dim(tables[[k+1]])))
    stop("All matrices must have identical dimensions!")
  }
  
  DNAME <- deparse(substitute(tables))
  
  if(compensated){
    tables = lapply(tables, function(X){
        X = X+(1/prod(dim(X)))
    })
  }

  null.table.marginal <- match.arg(null.table.marginal)
  # Get independent standard normal variables (E matrix) using Helmert tranfrom 
  if(null.table.marginal == "uniform"){
    EN <- get.e.mat.uni(tables)
    type <- "Null table marginal is uniform"
  }else{
    EN <- get.e.mat.obs(tables)
    type <- "Null table marginal is observed"
  }
  
  n <- EN$SamS

  if(all(n >= 0) && any(n > 0)) {

    K <- length(tables)

    EList <- EN$EList

    EP <- Reduce('+', EList) # Pooled E vectors of all conditions

    # Get the scaling coffecients:
    b <- sqrt(n) / sum(sqrt(n))

    # Difference of individual E vectors from E pooled:
    Q <- sapply(seq(K), function(k){
      return( EList[[k]] - b[k] * EP )
    })

    # Calculate covariance matrix:
    C <- cpc.cov.matrix(b)

    # Find the rank of the cov matrix
    R <- K - 1 # rankMatrix(C) 

    # Perform eigen-decomposition of covariance matrix:
    eig <- eigen(C)

    nonzero.eigenvalues <- eig$values[seq(R)]

    # Eigenvectors corresponding to non-zero eigenvalues:
    S <- eig$vectors[, seq(R), drop=FALSE]

    Z <- diag( 1 / sqrt( nonzero.eigenvalues ), nrow = R, ncol = R)

    # Calculating Mahalanobis distance squared from eigenvalue
    #   decomposition since rank is not full
    Stat <- sum( ( Q %*% S %*% Z ) ^ 2 )

    Df <- R * length(EP)  # Df <- R * ifelse(is.matrix(Q), nrow(Q), 1)
      
    P.val <- pchisq(Stat, Df, lower.tail = FALSE)

  } else {

    Stat = 0
    Df = 0
    P.val = 1

  }

  names(Stat) <- "X-squared"
  names(Df) <- "df"
  
  return(structure(list(
    statistic = Stat,
    parameter = Df,
    p.value = P.val,
    data.name = DNAME,
    method = paste("Sharma-Song Test for Second-Order Differential 
                   Contingency Tables","\n", type)),
    class = "htest"))
}


get.e.mat.obs <- function(tables)
{ 
  K <- length(tables)
  EList <- vector("list", length = K)
  n <- vector("numeric", length = K)

  for(k in seq(K)) {

    # Obtain row and column Helmert matrices
    rowSum <- rowSums(tables[[k]])  
    colSum <- colSums(tables[[k]])
    totalSum <- sum(rowSum)

    # Calculate expected table
    expec <- expected(tables[[k]])
    
    # Get normalized frequency table of sampled data
    A <- (tables[[k]] - expec) / sqrt(ifelse(expec == 0, 1, expec))
    
    # Correct the frequency for zero entries with both 
    #   a zero row sum and a zero column sum.
    #for(r in seq_along(rowSum)) {
    #   if(rowSum[r] != 0) next 
    #  for(c in seq_along(colSum)) {
    #     if(colSum[c] == 0) A[r, c] <- sqrt(totalSum)
    #  }
    # }

    # Vectorize E in column major 
    EList[[k]] <- get.e.mat(rowSum, colSum, totalSum, A)

    n[k] <- totalSum
  }
  return(list(SamS = n, EList = EList))
}



get.e.mat.uni <- function(tables)
{ 
  K <- length(tables)
  EList <- vector("list", length = K)
  n <- vector("numeric", length = K)
  
  for(k in seq(K)) {
    
    # Obtain row and column Helmert matrices
    rowSum <- rowSums(tables[[k]])  # (expec)
    colSum <- colSums(tables[[k]])  # (expec)
    totalSum <- sum(rowSum)
    
    R <- length(rowSum)
    S <- length(colSum)
    
    uniformSamp <- (totalSum)/(R*S)
    unimat = matrix(data = rep( uniformSamp,R*S), nrow = R, ncol = S)
    
    rowSum <- rowSums(unimat)  
    colSum <- colSums(unimat) 
    
    # Calculate expected table
    expec <- expected(tables[[k]])
    
    # Get normalized frequency table of sampled data
    A <- (tables[[k]] - expec) / sqrt(ifelse( uniformSamp == 0, 1,  uniformSamp))
    
    # Vectorize E in column major 
    EList[[k]] <- get.e.mat(rowSum, colSum, totalSum, A) 
    
    n[k] <- totalSum
  }
  return(list(SamS = n, EList = EList))
}

get.e.mat <- function(rowSum, colSum, totalSum, A){
  
  prdot <- rowSum / totalSum
  pcdot <- colSum / totalSum
  
  V <- helmert.matrix(prdot)
  W <- helmert.matrix(pcdot)
  
  
  # Transform normalized frequency matrix A to independent normal 
  #   variables by multiplying row and column helmert matrices
  E <- V %*% A %*% t(W)
  
  # Remove the first row and first column from E, which
  #   are alwarys zero. The other entries in E are random normal
  #   variables.
  E <- E[-1, -1]
  
  E <- as.vector(E)
  return(E)
  
}

expected <- function(table)
{
  rowSum = rowSums(table)
  colSum = colSums(table)
  totalSum = sum(rowSum)
  prod = outer(rowSum, colSum, "*")

  if(totalSum != 0) {
    Exp <- prod / totalSum
  } else {
    Exp <- prod
  }
  
  return (Exp)
}


helmert.matrix <- function(p)
{
  n <- length(p)
  L <- matrix(0, nrow=n, ncol=n)
  L[1, ] <- sqrt(p)

  sumpi.minus.1 <- p[1]

  if(n >= 2) {
    for(i in 2:n) {
      sumpi <- sumpi.minus.1 + p[i]

      # L[i,i] <- - sqrt(sumpi.minus.1  / sumpi) 
      L[i,i] <- - sqrt(
       ifelse(sumpi.minus.1 == 0,
              0,
              sumpi.minus.1  / sumpi) )

      # L[i,1:(i-1)] <- sqrt(p[i] * p[1:(i-1)] / (sumpi.minus.1 * sumpi))
      
      L[i,1:(i-1)] <- sqrt(
        ifelse( p[i] * p[1:(i-1)] == 0,
                0,
                p[i] * p[1:(i-1)] / (sumpi.minus.1 * sumpi)
              ) )

      sumpi.minus.1 <- sumpi
    }
  }
  return(L)
}


cpc.cov.matrix <- function(b)
{
  # bj > 0 and sum bj=1
  K <- length(b)
  JK <- matrix(1, nrow=K, ncol=K)
  covmat <- diag(K) - JK %*% diag(b) - diag(b) %*% JK + K * b %*% t(b)
  return(covmat)
}

